server/neptune/workflows/internal/deploy/revision/queue/worker.go

```diff
@@ -61,15 +72,66 @@ func (w *Worker) Work(ctx workflow.Context) {
 
 		msg := w.Queue.Pop()
 
-		ctx := workflow.WithValue(ctx, context.SHAKey, msg.Revision)
-		ctx = workflow.WithValue(ctx, context.DeploymentIDKey, msg.ID)
+		ctx := workflow.WithValue(ctx, internal_context.SHAKey, msg.Revision)
+		ctx = workflow.WithValue(ctx, internal_context.DeploymentIDKey, msg.ID)
 
-		err = w.TerraformWorkflowRunner.Run(ctx, msg)
+		isValid, err := w.isRequestRevisionValid(ctx, msg)
+		if err != nil {
+			logger.Error(ctx, fmt.Sprintf("Validating deploy request revision %s: %s", msg.Revision, err.Error()))
 
+			// TODO: Update the checkrun with relevant info
+			continue
+		}
+
+		if !isValid {
+			logger.Info(ctx, fmt.Sprintf("ID: %s Deploy Request Revision: %s is not valid", msg.ID, msg.Revision))
+
+			// TODO: Update the checkrun with relevant info
+			continue
+		}
+
+		err = w.TerraformWorkflowRunner.Run(ctx, msg)
 		if err != nil {
 			logger.Error(ctx, "failed to deploy revision, moving to next one")
 		}
+
+		err = w.persistLatestDeployment(ctx, msg)
+		if err != nil {
+			logger.Error(ctx, "failed to persist latest deploy job")
+		}
+	}
+}
+
+func (w *Worker) isRequestRevisionValid(ctx workflow.Context, deploymentInfo terraform.DeploymentInfo) (bool, error) {
+	// Fetch latest deployment
+	var resp activities.FetchLatestDeploymentResponse
+	err := workflow.ExecuteActivity(ctx, w.DbActivities.FetchLatestDeployment, activities.FetchLatestDeploymentRequest{
+		FullRepositoryName: w.Repo.GetFullName(),
+		RootName:           deploymentInfo.Root.Name,
+	}).Get(ctx, &resp)
+	if err != nil {
+		return false, errors.Wrap(err, "fetching latest deployment")
+	}
```

Comment:
Can't we just cache the latest deployment in memory (since temporal preserves state) and only fetch if we don't have this information (ie. when the workflow first starts).

For persistence, does it make sense to persist each item? Or should we just persist as part of shutdown process?

Reply:
Yeah I think it makes sense to only fetch the latest deployment when the workflow starts for the first time and hold it in memory.

For persistence, I think it depends on if we want the deployments data in s3 to be the single source of truth at all times. If we choose to persist on shutdown, the running workflow will only have the deployment data after it was run and the s3 bucket will only have the  deployment data for before it was run. 

In terms of reliability, if we are syncing up the deployment info at all times, we also have a backup s3 bucket for recovery. 

I think it makes sense to be a little precautionary and persist each deployment info after a workflow run since we might see some issues when we rollout that might leave the s3 bucket out of sync with the current state.

Reply:
I have changed the worker implementation to fetch on startup and persist after every workflow run!